src/wayland/input_method/qml.hpp

```diff
@@ -0,0 +1,160 @@
+	// QQmlComponent* mSurfaceComponent = nullptr;
+};
+
+/// Provides the ability to send text input to the compositor
+///
+/// A simple example that sends text to the currently focused input method:
+/// ```
+/// QSInputMethod {
+///   id: input_method
+/// }
+/// IpcHandler {
+///   target: "input_method"
+///   function hi(): void { input_method.sendString("hi"); }
+/// }
+/// ```
+/// We can now call the ipc handler and see that `hi` is printed to the terminal input.
+/// `$ qs ipc call input_method hi`
+class InputMethod: public QObject {
+	Q_OBJECT;
+	/// If a text input is focused
+	Q_PROPERTY(bool active READ isActive NOTIFY activeChanged);
+	/// Is false if another input method is already registered to the compositor, otherwise is true
+	Q_PROPERTY(bool hasInput READ hasInput NOTIFY hasInputChanged);
+	/// If the input method has grabbed the keyboard
+	Q_PROPERTY(bool hasKeyboard READ hasKeyboard NOTIFY hasKeyboardChanged);
+	/// The text around the where we will insert
+	Q_PROPERTY(QString surroundingText READ surroundingText NOTIFY surroundingTextChanged);
+	Q_PROPERTY(uint32_t surroundingTextCurosr READ surroundingTextCursor NOTIFY surroundingTextChanged);
+	Q_PROPERTY(uint32_t surroundingTextAnchor READ surroundingTextAnchor NOTIFY surroundingTextChanged);
+	/// The content_hint of the text input see https://wayland.app/protocols/text-input-unstable-v3#zwp_text_input_v3:enum:content_hint
+	Q_PROPERTY(QMLContentHint::Enum contentHint READ contentHint NOTIFY contentHintChanged);
+	/// The content_purpose of the text input see https://wayland.app/protocols/text-input-unstable-v3#zwp_text_input_v3:enum:content_purpose
+	Q_PROPERTY(QMLContentPurpose::Enum contentPurpose READ contentPurpose NOTIFY contentPurposeChanged);
+	/// Clear the preedit text when the keyboard grab is released
+	Q_PROPERTY(
+	    bool clearPreeditOnKeyboardRelease MEMBER mClearPreeditOnKeyboardRelease NOTIFY
+	        clearPreeditOnKeyboardReleaseChanged
+	);
+	/// The @@Keyboard$ that will handle the grabbed keyboard.
+	/// Use @@KeyboardTextEdit$ for most cases.
+	Q_PROPERTY(
+	    QQmlComponent* keyboard READ keyboardComponent WRITE setKeyboardComponent NOTIFY
+	        keyboardComponentChanged
+	);
+	Q_CLASSINFO("DefaultProperty", "keyboard");
+	QML_NAMED_ELEMENT(QSInputMethod);
+
+public:
+	explicit InputMethod(QObject* parent = nullptr);
+
+	/// Sends a string to the text input currently focused
+	Q_INVOKABLE void sendString(const QString& text);
+	/// Provides virtual text in the text input so the user can visualise what they write
+	/// This will override any previous preedit text
+	/// If `cursorBegin == cursorEnd == -1` the text input will not show a cursor
+	/// If `cursorBegin == cursorEnd == n` the text input will show a cursor at n
+	/// If `cursorBegin == n` and `cursorEnd == m` the text from n to m will be highlighted
+	Q_INVOKABLE void
+	sendPreeditString(const QString& text, int32_t cursorBegin = -1, int32_t cursorEnd = -1);
```

Comment:
The documentation generator doesn't properly support default args, so probably make two forms of this: `sendPreeditString(text, begin, end)`, `sendPreeditString(text)`